Added Caffeine Cache module and made it the default Cache

dusanstanojeviccs opened the pull request:
# Pull Request Checklist

* [x] Have you read [How to write the perfect pull request](https://github.com/blog/1943-how-to-write-the-perfect-pull-request)?
* [x] Have you read through the [contributor guidelines](https://www.playframework.com/contributing)?
* [x] Have you signed the [Lightbend CLA](https://www.lightbend.com/contribute/cla)?
* [x] Have you referenced any issues you're fixing using [commit message keywords](https://help.github.com/articles/closing-issues-using-keywords/)?
* [x] Have you added copyright headers to new files?
* [x] Have you checked that both Scala and Java APIs are updated?
* [x] Have you updated the documentation for both Scala and Java sections?
* [x] Have you added tests for any changed functionality?

# Helpful things

## Fixes

Fixes #7974 

## Purpose

This PR is adding Caffeine Cache module to Play and setting Caffeine Cache as the default cache provider.

## Background Context

I've mostly based this solution on previous Ehcache implementation.

## References

https://github.com/playframework/playframework/issues/7974

ben-manes commented on documentation/manual/working/javaGuide/main/cache/JavaCache.md:
```diff
@@ -1,29 +1,31 @@
 ## JCache Support
 
-Ehcache implements the [JSR 107](https://github.com/jsr107/jsr107spec) specification, also known as JCache, but Play does not bind `javax.caching.CacheManager` by default.  To bind `javax.caching.CacheManager` to the default provider, add the following to your dependencies list:
+Caffeine Cache implements the [JSR 107](https://github.com/jsr107/jsr107spec) specification, also known as JCache, but Play does not bind `javax.caching.CacheManager` by default.  To bind `javax.caching.CacheManager` to the default provider, add the following to your dependencies list:
```
Ehcache is a native JCache provider, since they were the main authors of the spec.

Caffeine isn't, since I consider the spec to be pretty bad. Instead it provides an optional implementation as an additional dependency. So you may need to clarify the additional caffeine-jcache dependency is required, too. Or you could consider it self-evident since jcache configuration is mostly non-standard.

dusanstanojeviccs replied:
Got it, changing the docs now will make a commit with more information and a reference to caffeine-jcache dependency.

ben-manes commented on documentation/manual/working/javaGuide/main/cache/JavaCache.md:
```diff
@@ -55,7 +57,7 @@ You can also supply a `Callable` that generates stores the value if no value is
 @[get-or-else](code/javaguide/cache/JavaCache.java)
 
-**Note**: `getOrElseUpdate` is not an atomic operation in Ehcache and is implemented as a `get` followed by computing the value from the `Callable`, then a `set`. This means it's possible for the value to be computed multiple times if multiple threads are calling `getOrElse` simultaneously.
+**Note**: `getOrElseUpdate` is not an atomic operation in Caffeine Cache and is implemented as a `get` followed by computing the value from the `Callable`, then a `set`. This means it's possible for the value to be computed multiple times if multiple threads are calling `getOrElse` simultaneously.
```
This is an unfortunate aspect of the Play! api. Caffeine does offer atomic operations, but not with a custom expiration passed through. Instead the `Expiry` is used to evaluate, since often the calls should prefer to use a `LoadingCache` rather than a `cache.get(key, func)`. This is the same problem with Ehcache and others, where that Play! api doesn't map neatly.

@gmethvin Is the lack of memoization a negative for Play! users? If so, you might want to implement this generically using lock striping for all providers to take advantage of. That's merely `Lock lock = locks[hash(key) mod locks.length]`, etc. There some minor optimization tricks, but would note.

dusanstanojeviccs replied:
If I understood that correctly get and set are not atomic if expiration is set and we should do locking it at the integration level?

ben-manes replied:
Yes, in that we don't provide a `computeIfAbsent(key, func, expiration)` type of call like Play! does. Neither does Ehcache, so both are using get-compute-put. There are atomic `computeIfAbsent(key, func)` type methods, but they call back into the Expiry to get the information. You would have to do some nasty thread-local to hack it where the TL is set, the Expiry queries it, and the TL is unset. I doubt that's really desirable.

dusanstanojeviccs replied:
Oh got it, so it would stay the same, right? Before there was just a note in the docs that the method was doing get-compute-put and everyone was ok with it, so I guess it would be ok if that stayed the same since Play! users would expect it.

ben-manes replied:
Yes, its just a quirk of Play's APIs rather than a missing feature. That API method does not map neatly to any caching library, and isn't worth providing ad hoc just for Play when it is fairly ugly (when compared to the preferred). Since this is an issue across all providers, the best fix might be to add the feature into Play's abstraction. Using lock striping is cheap and simple code, so a suggestion for @gmethvin to consider someday.

gmethvin replied:
Yes, this is an unfortunate design quirk of the Play cache. There has also been discussion of allowing a configurable default for the methods when an explicit expiration is not passed, which would probably make more sense.

@dusanstanojeviccs I think for the purposes of this PR, the goal should be parity with the current ehcache-based implementation, so this disclaimer would apply to both caffeine and ehcache. If you have interest in implementing lock striping to solve the problem as @ben-manes described, that would be a nice addition, but it could also be done in a future PR.

ben-manes replied:
Yes, I am only putting this on @gmethvin radar and not asking for it in this PR. Ignore this distraction.

dusanstanojeviccs replied:
I can probably do this and I like the locks[hash(key) mod locks.length], but what should the size of locks be, should it be scaled up and down depending on the number of keys currently in the array (like an arraylist) in sync(locks) blocks or should it just have a constant size set up front in the config file?

ben-manes replied:
It would be in a subsequent PR.

Generally you size ad hoc, since any decent number will offer enough concurrency. If you try to be fancy, you can base it on `Runtime.getRuntime().availableProcessors()` but it probably won't help. A setting like 2048 is plentiful and perhaps larger than necessary. Generally you want to minimize collisions but since writes are rarer, the likelihood of overlap isn't high. You would use a power-of-two in order to use the cheaper modulus trick of `N & (N - 1)`. There always reaches a point where increasing the number of stripes stops being beneficial and a redesign is required instead.

The stripe can be an immutable Object[] array with unique instances. Then one would use double-checked locking to avoid synchronizing in the happy-path of the entry being present. This way calls for the same key will collide and be blocked, with a high probability for distinct keys compute concurrently.

```java
V value = cache.getIfPresent(key);
if (value != null) {
  return value;
}
Object lock = locks[hash(key) & MASK];
synchronized (lock) {
  value = cache.getIfPresent(key);
  if (value != null) {
    value = // compute
    cache.put(key, value, expiration);
  }
  return value;
}
```

A supplementary hash function, like murmur3 or wang-jenkins, would help the distribution if `key.hashCode()` was of poor quality (which I think String is). For example the spreader in the original `ConcurrentHashMap` (based on fixed stripes vs dynamic tree-bins in JDK8) was

```java
private static int hash(int h) {
  // Spread bits to regularize both segment and index locations,
  // using variant of single-word Wang/Jenkins hash.
  h += (h <<  15) ^ 0xffffcd7d;
  h ^= (h >>> 10);
  h += (h <<   3);
  h ^= (h >>>  6);
  h += (h <<   2) + (h << 14);
  return h ^ (h >>> 16);
}
```

I would imagine a striped CacheApi would be a trait or decorator for reuse - whatever is idiomatic these days. The underlying cache would assumed to be concurrent, so only `set` calls would need to use the lock to compute with.

dusanstanojeviccs replied:
@ben-manes Sorry I have not responded sooner, I was traveling and then had to catch up on work. Wow! That is beautiful and makes total sense, the whole N & (N - 1) is brilliant! I will definitely do this in a future PR. Thank you for the pointers.

ben-manes commented on documentation/manual/working/javaGuide/main/cache/JavaCache.md:
```diff
@@ -71,26 +73,26 @@ Note that the [SyncCacheApi](api/java/play/cache/SyncCacheApi.html) has the same
+If you want to pass custom configuration data for Caffeine Cache you can do so by specifying:
 
-    play.cache.createBoundCaches = false
+    play.cache.caffeine.spec = {}
```
I think you changed from the spec string to the parser?

dusanstanojeviccs replied:
I did, but I named the configuration value 'play.cache.caffeine.spec'. Would it be clearer for Caffeine users if I changed it to 'play.cache.caffeine.parser'?

ben-manes replied:
Oh, nevermind then. I thought this was the unparsed spec string. It might be good to see an example of how a named cache works (just here for me to grok it). I'd have thought it would be `play.cache.caffeine.<name>`.

dusanstanojeviccs replied:
I've looked at the current ehcache.xml example and it looked like it was never bound to a cache name so I figured that was not part of the spec before. But to be honest that makes a lot of sense. Doing it now.

dusanstanojeviccs replied:
Done, running tests now.

dusanstanojeviccs replied:
So this is how the spec would look like now:

caffeine.spec = {
  customCacheName = {
    maximum-size: 200
  }
  play = {
    maximum-size: 243
  }
}

'play' is the default cache and 'customCacheName' is a custom caches name. That is the expected way it should work, right?

ben-manes replied:
Why not drop the `spec`?

```
play.cache.caffeine {
  play {
    maximum-size: 200
  }
  custom {
    ...
  }
}
```

The Caffeine JCache adapter uses this type of [configuration](https://github.com/ben-manes/caffeine/blob/master/jcache/src/main/resources/reference.conf). but I don't know if it is considered idiomatic. Hopefully the core devs here have an opinion.

dusanstanojeviccs replied:
No good reason to keep the spec so I dropped it. I have not seen any modules that do this so I am not sure if it is idiomatic either. I would imagine it would be fine as long as it got documented in the official docs but let's hold off on it until we get an opinion from one of the core devs. :)

gmethvin replied:
How about something like:

```
play.cache.caffeine {
  defaults {
    maximumSize = 200
    ...
  }
  caches {
    play = {}
    custom-cache = {}
  }
}
```

This is similar to what we do on database configuration: https://github.com/playframework/playframework/blob/master/framework/src/play-jdbc/src/main/resources/reference.conf#L28.

We can also include the reference.conf directly in the page like we do on other pages, e.g. https://github.com/playframework/playframework/blob/2.6.x/documentation/manual/working/commonGuide/configuration/SettingsAkkaHttp.md

dusanstanojeviccs replied:
@gmethvin Looks good, I've added the code for this in my last commit. So to be clear defaults is used if "play.cache.caffeine.caches" does not contain the cache name as a child.

gmethvin replied:
The configuration for each cache would be layered on top of the default. The presence of the `play` key would indicate that a cache called `play` should be created, and and the values in that object would override the default values with the same name.

ben-manes commented on framework/src/play-caffeine-cache/src/main/java/play/cache/caffeine/CaffeineParser.java:
```diff
@@ -0,0 +1,116 @@
+    private void parse(String key) {
+        switch (key) {
+            case "initial-capacity":
+                System.out.println("initial-capacity:" + config.getInt(key));
```
Does Play! not have a logger?

dusanstanojeviccs replied:
It does, I've put this for checking if everything works, I just personally like using System.out because I see no advantage of logger when testing... I've removed it now.

ben-manes commented on framework/src/play-caffeine-cache/src/main/java/play/cache/caffeine/CaffeineParser.java:
```diff
@@ -0,0 +1,116 @@
+                System.out.println("maximum-weight:" + config.getMemorySize(key));
+                cacheBuilder.maximumWeight(config.getMemorySize(key).toBytes());
+                break;
+            case "expire-after-access":
```
Since variable expiration is required for the play API, these fixed versions cannot be used simultaneously. So you should remove.

dusanstanojeviccs replied:
Got it, I've removed: expire-after-access, expire-after-write, refreshAfterWrite.

ben-manes commented on framework/src/play-caffeine-cache/src/main/scala/play/api/cache/caffeine/CaffeineCacheApi.scala:
```diff
@@ -0,0 +1,247 @@
+        val seconds = finite.toSeconds
+        if (seconds <= 0) {
+          cache.policy().expireVariably().get().put(key, value, 1, TimeUnit.SECONDS)
+        } else if (seconds > Int.MaxValue) {
```
Since the `duration` is a `long` parameter, you can probably drop this?

dusanstanojeviccs replied:
Got it, because this was only needed before because the expiration was int, now it's pointless, good catch.

ben-manes commented:
What is Play's approach to exporting statistics?

Ehcache does it using JMX, but it is more framework-oriented than Caffeine. For us, we expose stats as pull (`Cache#stats()`) or push (`recordStats(collector)`). Then users can bridge into their preferred system. Would users inject in the named cache(s) to do that? It would be good to have a short note on how to collect the stats.

gmethvin commented:
@dusanstanojeviccs Haven't had a chance to fully review, but this should be another module, in addition to the existing ehcache module. It should be no problem to have both modules available to use. Perhaps in the future we can remove the ehcache module but I think it would be too disruptive to do it immediately in the next release.

dusanstanojeviccs commented:
@gmethvin Got it, I think I removed it when I started working on this because I did not have a good grasp of how the whole Play build system with sbt is organized, but I have brought it back and all  'sbt tests' have passed :)

I have addressed the coding issues.

Now I am working on improving the docs for Play 2.7 migration and adding/correcting a few things about caffeine and jcache. Hopefully, the docs will be done tonight.

ben-manes commented on framework/src/play-caffeine-cache/src/main/java/play/cache/caffeine/CaffeineCacheManager.java:
```diff
@@ -11,25 +11,37 @@
+        Caffeine cacheBuilder;
+
+        // we create the cache builder
+        // depending on weather the config for
```
whether

dusanstanojeviccs replied:
yup

dusanstanojeviccs commented:
@marcospereira Loved the doc suggestions. I will do a few more doc changes tonight, but I think I'll do both of your requested changes tomorrow.

gmethvin commented on framework/build.sbt:
```diff
@@ -218,9 +218,9 @@ lazy val PlayAhcWsProject = PlayCrossBuiltProject("Play-AHC-WS", "play-ahc-ws")
-  ).dependsOn(PlayWsProject, PlayEhcacheProject % "test")
-  .dependsOn(PlaySpecs2Project % "test")
-  .dependsOn(PlayTestProject % "test->test")
+  ).dependsOn(PlayWsProject, PlayCaffeineCacheProject)
```
Does this need to be a compile dependency rather than just a test dependency?

dusanstanojeviccs replied:
If it was test before it should be test now since I did not change the Ahc module in any way, fixed and committed

gmethvin commented on framework/src/play-caffeine-cache/src/main/java/play/cache/caffeine/CaffeineParser.java:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    @SuppressWarnings("fallthrough")
+    private void parse(String key) {
```
I would prefer if we explicitly got the configuration elements we need.

These should all be documented in `reference.conf`. Ideally our `defaults` should look something like:
```
defaults = {
  initial-capacity = null
  weak-keys = null
  maximum-weight = null
  weak-keys = false
  soft-values = false
  record-stats = false
}
```
In this case `null` can mean to use caffeine's default value. If Play wants to override a value by default, then it's as easy as updating the `reference.conf`.

Then we can explicitly get each of the keys. I don't really mind if you write this in Java but it would probably be simpler to write it in Scala using the play `Configuration` helper:
```scala
configuration.get[Option[Int]]("initial-capacity") foreach cacheBuilder.initialCapacity
configuration.get[Option[Long]]("maximum-size") foreach cacheBuilder.maximumSize
if (configuration.get[Boolean]("soft-values")) cacheBuilder.softValues()
```
and so on.

dusanstanojeviccs replied:
Cool, fallthrough has been implemented, I've kept the file in Java and just put 3 if statements around it for null to not set the values. It should be pretty easy to understand.

ben-manes replied:
maximum-weight requires a Weigher, but I don’t think there is a current way to pass in an instance. A class literal with a Guice injector query would be a sad solution, though. So I don’t think this feature can be exposed without some additional work.

dusanstanojeviccs replied:
Got it, so I've removed maximum-weight, is that the right thing to do for now? If the feature becomes exposed we can always add it to a future PR.

gmethvin replied:
I'm not sure I buy that fallthrough is the right way to do this. Even if you don't want to use the scala `Configuration` wrapper, you can use:

```java
private void nullable(String key, Consumer<String> ifNotNull) {
  if (!config.getIsNull(key)) {
    ifNotNull(key);
  }
}
```
then getting nullable keys is as easy as:
```java
nullable("initial-capacity", key -> cacheBuilder.initialCapacity(config.getInt(key)))
nullable("maximum-size", key -> cacheBuilder.maximumSize(config.getLong(key)))
```
If we have default values in `reference.conf` we don't need to worry about the case where the key does not exist.

And I don't think this code here works for null values.

dusanstanojeviccs replied:
I might not be understanding what you mean. Nulls are not connected to the fallthrough and are handled on line 64 in CaffeineParser, since there are 2 places they are called the whole nullable method seems like an overkill as compared to 2 if statements. I might be wrong but let me know if I misunderstood something.

ben-manes replied:
Probably doesn't help, but I use fall-through and a default template in a similar way in my JCache adapter. It has to deal with some merging logic, too.

https://github.com/ben-manes/caffeine/blob/master/jcache/src/main/java/com/github/benmanes/caffeine/jcache/configuration/TypesafeConfigurator.java

gmethvin replied:
Oh, I think you made the changes to handle nulls after I commented here. Still, I don't understand why the `switch` statement is necessary for this at all.

gmethvin commented on framework/src/play-caffeine-cache/src/main/scala/play/api/cache/caffeine/CaffeineCacheApi.scala:
```diff
@@ -0,0 +1,247 @@
+ * CaffeineCache components for compile time injection
+ */
+trait CaffeineCacheComponents {
+  def environment: Environment
```
It doesn't look like you need `environment` or `applicationLifecycle` in here at all.

dusanstanojeviccs replied:
Yup, it's removed :)

gmethvin commented on framework/src/play-caffeine-cache/src/main/scala/play/api/cache/caffeine/CaffeineCacheApi.scala:
```diff
@@ -0,0 +1,247 @@
+  /**
+   * Use this to create with the given name.
+   */
+  def cacheApi(name: String, create: Boolean = true): AsyncCacheApi = {
```
Is the `create` needed here? We don't need to worry about having the exact same API as `EhCacheComponents`.

dusanstanojeviccs replied:
It's not used and since we dont have to keep the same API it is removed

gmethvin commented on framework/src/play-caffeine-cache/src/main/java/play/cache/caffeine/CaffeineConfigConstants.java:
```diff
@@ -0,0 +1,15 @@
+ */
+package play.cache.caffeine;
+
+public class CaffeineConfigConstants {
```
Reading your code I actually don't find it useful to have these in global string constants. Generally most of these need to be referenced in exactly one place besides the configuration.

dusanstanojeviccs replied:
I agree with you on this one, I was planning to remove them. I usually don't like having string literals in the code but they seem to be better for readability when it comes to config keys...

gmethvin commented on framework/src/play-caffeine-cache/src/main/scala/play/api/cache/caffeine/CaffeineCacheApi.scala:
```diff
@@ -0,0 +1,247 @@
+    new CaffeineCacheApi(NamedCaffeineCacheProvider.getNamedCache(name, caffeineCacheManager, configuration))(ec)
+  }
+
+  lazy val defaultCacheApi: AsyncCacheApi = cacheApi(PLAY_CACHE_CAFFEINE_DEFAULT_NAME)
```
I think this should use the default cache name, i.e. the value of the `play.cache.defaultCache` configuration.

dusanstanojeviccs replied:
Yes, that is a good catch, interestingly this was always passing "play" in the EhCache implementation

gmethvin commented on framework/src/play-caffeine-cache/src/main/java/play/cache/caffeine/CaffeineCacheComponents.java:
```diff
@@ -0,0 +1,65 @@
+ * }
+ * </pre>
+ */
+public interface CaffeineCacheComponents extends ConfigurationComponents, AkkaComponents {
```
Same comments as in the scala components.

dusanstanojeviccs replied:
Both are done

dusanstanojeviccs commented on framework/src/play-caffeine-cache/src/main/java/play/cache/caffeine/CaffeineCacheManager.java:
```diff
@@ -0,0 +1,59 @@
+
+            if (config.hasPath(pathKey)) {
+                cacheBuilder = CaffeineParser.from(config.getConfig(pathKey)).expireAfter(defaultExpiry);
+            } else if (config.hasPath(PLAY_CACHE_CAFFEINE_DEFAULTS)) {
```
YES! I did not know withFallback existed, it made the code so much simpler. Thank you!

ben-manes commented:
I think so. There are a few similar features, like removal listener, that
could go into a future PR when users actually request them. Minimal seems
wise for now.

On Tue, Dec 5, 2017 at 10:31 PM dusanstanojeviccs <notifications@github.com>
wrote:

> *@dusanstanojeviccs* commented on this pull request.
> ------------------------------
>
> In
> framework/src/play-caffeine-cache/src/main/java/play/cache/caffeine/CaffeineParser.java
> <https://github.com/playframework/playframework/pull/8025#discussion_r155155644>
> :
>
> > +    private CaffeineParser(Config config) {
> +        this.cacheBuilder = Caffeine.newBuilder();
> +        this.config = Objects.requireNonNull(config);
> +    }
> +
> +    /** Returns a configured {@link Caffeine} cache builder. */
> +    public static Caffeine<Object, Object> from(Config config) {
> +        CaffeineParser parser = new CaffeineParser(config);
> +
> +        config.entrySet().stream().map(Map.Entry::getKey).forEach(parser::parse);
> +
> +        return parser.cacheBuilder;
> +    }
> +
> +    @SuppressWarnings("fallthrough")
> +    private void parse(String key) {
>
> Got it, so I've removed maximum-weight, is that the right thing to do for
> now? If the feature becomes exposed we can always add it to a future PR.
>
> —
> You are receiving this because you were mentioned.
>
> Reply to this email directly, view it on GitHub
> <https://github.com/playframework/playframework/pull/8025#discussion_r155155644>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXG9oeed-OEucbBjYCk0kx4I7Z4Grs0ks5s9jTJgaJpZM4Qlawz>
> .
>

gmethvin commented on framework/src/play-caffeine-cache/src/main/resources/reference.conf:
```diff
@@ -0,0 +1,20 @@
+  cache {
+    # Data that should be used to configure the cache
+    caffeine {
+      defaults {}
```
Can you add the default values to reference.conf? It should look something like:

```
defaults {
  initialCapacity = null
  maximumSize = null
  weakKeys = false
  softValues = false
  recordStats = false
}
```
Even if the value is `null` meaning the caffeine default, it's useful to have the settings documented here.

I was also thinking we should use camelCase here since that's consistent with what caffeine uses and camelCase is used in other places in this configuration. What do you think?

ben-manes replied:
According to the HOCON spec,
> Config keys are encouraged to be hyphen-separated rather than camelCase.

https://github.com/lightbend/config/blob/master/HOCON.md#hyphen-separated-vs-camelcase

dusanstanojeviccs replied:
I did this with hyphen separated values:

```
    caffeine {
      defaults {
        initial-capacity = null
        weak-keys = null
        weak-keys = false
        soft-values = false
        record-stats = false
      }
      caches {}
    }
```

gmethvin commented:
For stats, I think the most basic implementation would be for the user to inject the cache instance and call `stats()`. Eventually I think we'd want to let the user bind their own instance of `StatsCollector` but that requires a bit more thinking so it's probably best to do that in another PR.

dusanstanojeviccs commented:
@marcospereira Hey, I've added the docs, moved the old ehcache docs for people who want to use it from /cache to /ehcache (both scala and java) added a paragraph to migrations to 2.7 and I've made sure it works.

@gmethvin @marcospereira I had one more question and I'm hoping you can help me figure it out, in the PlayImport.scala I've added val caffeine = component("play-caffeine-cache") but I did not change the value of cache because it is deprecated. So I guess my question is if we are moving the default implementation to caffeine should we change this value?

gmethvin commented:
@dusanstanojeviccs No, we should not change the `cache` dependency. If anything we could remove it for 2.7, but that could probably be decided in another PR. The point of separating `cacheApi` and `ehcache` was to make the cache implementation explicit to the user. Now when we introduce the caffeine implementation it will be obvious.

gmethvin commented on documentation/manual/working/javaGuide/main/cache/JavaCache.md:
```diff
@@ -1,29 +1,33 @@
 For any data stored in the cache, a regeneration strategy needs to be put in place in case the data goes missing. This philosophy is one of the fundamentals behind Play, and is different from Java EE, where the session is expected to retain values throughout its lifetime.
 
-The default implementation of the cache API uses [Ehcache](http://www.ehcache.org/).
+The default implementation of the cache API uses [Caffeine Cache](https://github.com/ben-manes/caffeine/).
```
I notice there is another page for ehcache, so I think we should link it here, but point out that we recommend using caffeine if you're looking for a local cache.

gmethvin commented on documentation/manual/releases/release27/migration27/CacheMigration27.md:
````diff
@@ -0,0 +1,26 @@
+)
+```
+
+You can find the full documentation for using EhCache with Java here [[ehcache documentation for java|JavaEhCache]]. If you are using Scala the documentation is [[here|ScalaEhCache]].
````
Usually it's nicest to integrate this into the wording like: `Read the documentation for the [[Java Ehcache API|JavaEhCache]] and [[Scala Ehcache API|ScalaEhCache]] to learn more about configuring Ehcache with Play.`

gmethvin commented on documentation/manual/releases/release27/migration27/CacheMigration27.md:
```diff
@@ -0,0 +1,26 @@
+# Cache Migration
```
I actually think this would be better as a Highlights section. This is really more like a new feature.

gmethvin commented on framework/src/play-caffeine-cache/src/main/java/play/cache/caffeine/CaffeineParser.java:
```diff
@@ -0,0 +1,89 @@
+ * <ul>
+ *   <li>{@code initial-capacity=[integer]}: sets {@link Caffeine#initialCapacity}.
+ *   <li>{@code maximum-size=[long]}: sets {@link Caffeine#maximumSize}.
+ *   <li>{@code maximum-weight=[memory-size]}: sets {@link Caffeine#maximumWeight}.
```
`maximum-weight` and some other ones aren't actually used here.

gmethvin commented on documentation/manual/working/scalaGuide/main/ehcache/ScalaEhCache.md:
```diff
@@ -0,0 +1,143 @@
+<!--- Copyright (C) 2009-2017 Lightbend Inc. <https://www.lightbend.com> -->
+# The Play cache API
```
I feel like these pages need to be restructured. Right now it looks like this is just a copy of the other page that describes both the cache API and now the caffeine implementation, so it feels redundant.

It seems like we could just have a single page and have subheadings for when the behavior differs, e.g. you can use `ehcache.xml` for ehcache and config for caffeine.

ben-manes commented:
Happy new year. Can we revisit this?

dusanstanojeviccs commented:
@ben-manes Happy new year! I'm so sorry this has dragged on for so long, I was super busy with work for the past few months. I am planning to take some time off this weekend and finish up the docs fully.

dusanstanojeviccs commented:
@ben-manes Just committed the changes to the documentation.

@gmethvin Hey, you were absolutely right, the documentation was mostly implementation agnostic and the differences between caffeine and ehcache where few, so I've grouped it together into one file and I've used subheadings for the implementation differences. Should be much more readable now.

@marcospereira Hey, I've added documentation to Highlights27 that explains that we now have the implementation for Caffeine and that it is the recommended cache implementation. Documentation now contains both Caffeine and EhCache information so the users can choose. Let me know if there is anything else.

ben-manes commented on documentation/manual/working/javaGuide/main/cache/JavaCache.md:
```diff
@@ -1,36 +1,57 @@
+Caffeine Cache does not natively implement the [JSR 107](https://github.com/jsr107/jsr107spec) specification, also known as JCache. If you want to use JCache with Caffeine you can use [caffeine-jcache](https://github.com/ben-manes/caffeine/wiki/JCache) dependency.
+
+### JCache support with EhCache
+Ehcache implements the [JSR 107](https://github.com/jsr107/jsr107spec) specification, also known as JCache, by default.
```
Actually only in v3 is that true, but Play! uses v2 instead. There is an unmaintained adapter, though. In v3 it doesn’t offer expiration apis compatible with Play’s so its kind of a mess.

dusanstanojeviccs replied:
Hmm so what would be your recommendation for this part of the documentation, point to the adapter or just say that there is no maintained JCache implementation with Caffeine?

ben-manes replied:
Sorry, I meant it is unmaintained for ehcache v2
https://github.com/ehcache/ehcache-jcache

You can reference it or users could use the v3 play adapter
https://github.com/ehcache/play-jcache

It might actually support expiration and that I misremembered

PromanSEW commented on documentation/manual/working/javaGuide/main/cache/JavaCache.md:
```diff
@@ -7,16 +7,28 @@ Caching data is a typical optimization in modern applications, and so Play provi
 For any data stored in the cache, a regeneration strategy needs to be put in place in case the data goes missing. This philosophy is one of the fundamentals behind Play, and is different from Java EE, where the session is expected to retain values throughout its lifetime.
 
-The default implementation of the cache API uses [Caffeine Cache](https://github.com/ben-manes/caffeine/).
+Implmentations of both [Caffeine](https://github.com/ben-manes/caffeine/) and [EhCache](http://www.ehcache.org) are provided, using the Caffeine implementation is recommended.
```
Typo in the first word: Impl**e**mentation

PromanSEW commented on documentation/manual/working/javaGuide/main/cache/code/javaguide/ehcache/JavaEhCache.java:
```diff
@@ -0,0 +1,119 @@
+        CompletionStage<Done> result = cache.set("item.key", frontPageNews);
+        //#simple-set
+        block(result);
+        //#time-set
```
This line must be 2 lines below as in `cache/JavaCache.java`

PromanSEW commented on documentation/manual/working/scalaGuide/main/cache/ScalaCache.md:
```diff
@@ -7,32 +7,51 @@ Caching data is a typical optimization in modern applications, and so Play provi
 For any data stored in the cache, a regeneration strategy needs to be put in place in case the data goes missing. This philosophy is one of the fundamentals behind Play, and is different from Java EE, where the session is expected to retain values throughout its lifetime.
 
-The default implementation of the Cache API uses [Ehcache](http://ehcache.org/).
+Implmentations of both [Caffeine](https://github.com/ben-manes/caffeine/) and [EhCache](http://www.ehcache.org) are provided, using the Caffeine implementation is recommended.
```
Typo in the first word: Implementation

gmethvin commented on documentation/manual/working/scalaGuide/main/cache/ScalaCache.md:
```diff
@@ -7,32 +7,51 @@ Caching data is a typical optimization in modern applications, and so Play provi
 
+### JCache support with Caffeine
+
+Caffeine Cache does not natively implement the [JSR 107](https://github.com/jsr107/jsr107spec) specification, also known as JCache. If you want to use JCache with Caffeine you can use [caffeine-jcache](https://github.com/ben-manes/caffeine/wiki/JCache) dependency.
```
So actually I think the `jcache` module we need is already there in `playCaffeineDeps` so it shouldn't need to be added, right? And that documentation page just talks about configuring the JCache support for Caffeine, not how to add the dependency.

There is another confusing thing here: as far as I can tell, the JCache support for Caffeine would create a separate cache based on a separate configuration (which also uses typesafe config). @ben-manes what do you think is the best way to unify these two so we can have a common configuration?

ben-manes replied:
I don't think unifying is a great option. All JCache implementations need to provide their own configuration support and each do it differently. Most appear to prefer XML, it seems. It would seem that you might want a generic integration strategy (e.g. Play JCache adapter) that fetches from the `CacheManager` than to couple to an implementation.

Caffeine's JCache will likely be written in v3 to simplify it. The reason is that variable expiration was a late feature, since doing it correctly is hard, and most JCache implementations use lazy expiration by requiring a maximum size. This behavior was emulated in the JCache layer, and later a timerwheel-based approach was implemented in the core library. This will result in a simpler config and implementation, so I'd like the option to not tie it deeply with Play's.

Also, why does Play appear to recommend JCache? If you've looked into it, its actually an awful specification. Many of its features are incompatible with each other, it was rejected by JEE multiple times (most recently by JEE-8), and has a poor user experience. It is useful for integrations, like Hibernate, but otherwise I consider it to be a step backwards. I certainly wouldn't use it in any of my code.

gmethvin replied:
Perhaps this documentation needs to be reworded since we should definitely not be recommending JCache.

I'm not actually sure why we introduced JCache support in the first place. Probably to allow integration with something else. @wsargent can you provide any background? I think it was used in the WS API to provide a common API for caching requests.

Maybe the rewording should be something like "We do not recommend using JCache in your own code, but it may be useful for integration with other libraries. Both Caffeine and Ehcache provide JCache implementations."

dusanstanojeviccs replied:
@gmethvin I have added those sentences to the "JCache Support" header.

wsargent replied:
> I'm not actually sure why we introduced JCache support in the first place. Probably to allow integration with something else. @wsargent can you provide any background? I think it was used in the WS API to provide a common API for caching requests.

The theory was that we may add Caffeine, but people would still be using Ehcache (given how long its stuck around) and so JCache is the lowest common denominator API between them, same as reactive streams is the base API for akka streams.

dusanstanojeviccs replied:
@wsargent You are absolutely right. From what I've seen JCache API is used for ws caching only. Regular caching is implemented directly on top of plays CacheApi (lowest denominator) interface so it has no dependencies on JCache. I have removed the JCache section from the caching page but have not touched the ws documentation, ws still uses JCache but does a really good job of explaining how to use it so I figured better not touch those docs.

dusanstanojeviccs commented on framework/project/Dependencies.scala:
```diff
@@ -275,11 +279,15 @@ object Dependencies {
     "com.typesafe.play" % "shaded-oauth" % playWsStandaloneVersion,
-    "com.github.ben-manes.caffeine" % "jcache" % caffeineVersion % Test
+    "com.github.ben-manes.caffeine" % "jcache" % caffeineVersion % Test,
+    "net.sf.ehcache" % "ehcache" % ehcacheVersion,
```
Hey, thanks for pointing this out, the file OptionalAhcHttpCacheProviderSpec.scala has a test written for using ehcache with jcache, I've mirrored this into one more test like this:

```
val settings = Map(
        "play.ws.cache.enabled" -> "true",
        "play.ws.cache.cachingProviderName" -> classOf[CaffeineCachingProvider].getName,
        "play.ws.cache.cacheManagerResource" -> "caffeine.conf"
 )
```

So the whole ahc is now dependent on both caffeine and ehcache for this test to run. I am not sure if we want to change the ahc in this PR, if we want to do that let me know in which way and I'll adjust it.

gmethvin replied:
If you just need it for tests then it should be:

```scala
"net.sf.ehcache" % "ehcache" % ehcacheVersion % Test
```

test dependencies are generally fine, but I'd prefer not to add runtime dependencies to play-ahc-ws.

dusanstanojeviccs replied:
Oh got it, that makes sense, thanks for pointing that out.

dusanstanojeviccs commented on framework/src/play-caffeine-cache/src/main/resources/reference.conf:
```diff
@@ -0,0 +1,26 @@
+
+  cache {
+    # Data that should be used to configure the cache
+    caffeine {
```
Written a new test for this, the only way I could see for testing this is to make sure that the proper cache builder is set up depending on the cache name (because the name is passed to it from the CaffeineCacheManager.getCache).

dusanstanojeviccs commented on documentation/manual/working/scalaGuide/main/cache/ScalaCache.md:
```diff
@@ -73,26 +92,43 @@ Note that the [SyncCacheApi](api/scala/play/api/cache/SyncCacheApi.html) has the
 
     play.cache.bindCaches = ["db-cache", "user-cache", "session-cache"]
 
-By default, Play will try to create these caches for you. If you would like to define them yourself in `ehcache.xml`, you can set:
```
I've added a paragraph for EhCache configuration with named cache:

<img width="693" alt="screen shot 2018-02-14 at 6 58 19 pm" src="https://user-images.githubusercontent.com/15340338/36238699-17c5e7cc-11b9-11e8-9169-a9fa10b6462c.png">

gmethvin commented on documentation/manual/releases/release27/Highlights27.md:
```diff
@@ -24,3 +24,11 @@ public class MyForm {
+
+## Support for Caffeine
+
+Play now offeres an implementation of caching with Caffeine. Caffeine is the recomended cache implementation for Play users.
```
"Play now offers a CacheApi implementation based on [Caffeine](https://github.com/ben-manes/caffeine/)." (with the link)

gmethvin commented on documentation/manual/releases/release27/Highlights27.md:
```diff
@@ -24,3 +24,11 @@ public class MyForm {
+
+Play now offeres an implementation of caching with Caffeine. Caffeine is the recomended cache implementation for Play users.
+
+To migrate from EhCache to Caffeine you will have to remove `ehcache` from your dependencies and replace it with `caffeine` and you will have to change your caching configuration.
```
Run-on sentence. Could be rewritten as: "To migrate from EhCache to Caffeine you will have to remove `ehcache` from your dependencies and replace it with `caffeine`. To customize the settings from the defaults you will also need to update the configuration in `application.conf` as explained in the documentation."

ben-manes commented:
Thanks @dusanstanojeviccs!

@gmethvin are there remaining items for review?

gmethvin commented on framework/src/play-caffeine-cache/src/test/scala/play/api/cache/JavaCacheApiSpec.scala:
```diff
@@ -0,0 +1,142 @@
+      val cacheApi = app.injector.instanceOf[JavaAsyncCacheApi]
+      Await.result(cacheApi.set("foo", "bar", 1 /* second */ ).toScala, 1.second)
+
+      Thread.sleep(2.seconds.toMillis)
```
Do we need these `Thread.sleep` calls?

dusanstanojeviccs replied:
I think that particular test sets expiration to 1 second, waits 2 seconds and check if the value really got cleared from the cache.

marcospereira replied:
We already have some of these in our specs. It could be good to move them later to use something like [specs2 `eventually`](https://etorreborre.github.io/specs2/guide/SPECS2-4.0.2/org.specs2.guide.Matchers.html#derive-matchers).

But this may be out of the scope of this PR.

gmethvin commented on documentation/manual/working/javaGuide/main/cache/JavaCache.md:
```diff
@@ -1,35 +1,41 @@
 For any data stored in the cache, a regeneration strategy needs to be put in place in case the data goes missing. This philosophy is one of the fundamentals behind Play, and is different from Java EE, where the session is expected to retain values throughout its lifetime.
 
-The default implementation of the cache API uses [Ehcache](http://www.ehcache.org/).
+Play provides a CacheApi implementation based on [Caffeine](https://github.com/ben-manes/caffeine/) and a legacy implementation based on [Ehcache 2.x](http://www.ehcache.org). For in-process caching Caffeine is typically the best choice. If you need distributed caching, there are third-party plugins for memcached and redis.
```
You might want to link to https://github.com/mumoshu/play2-memcached and https://github.com/KarelCemus/play-redis.

marcospereira replied:
Or maybe create a "Cache" section on `ModuleDirectory.md` and link to there.

marcospereira commented on framework/src/play-caffeine-cache/src/main/java/play/cache/caffeine/CaffeineDefaultExpiry.java:
```diff
@@ -0,0 +1,25 @@
+
+import com.github.benmanes.caffeine.cache.Expiry;
+
+import javax.annotation.Nonnull;
```
Nice. We may want to start using these annotations more widely. :-)

marcospereira commented on framework/src/play-caffeine-cache/src/test/scala/play/api/cache/JavaCacheApiSpec.scala:
```diff
@@ -0,0 +1,142 @@
+  "Java AsyncCacheApi" should {
+    "set cache values" in new WithApplication {
+      val cacheApi = app.injector.instanceOf[JavaAsyncCacheApi]
+      Await.result(cacheApi.set("foo", "bar").toScala, 1.second)
```
You can just use `await(cacheApi("foo", "bar").toScala)` instead, which already has a default timeout and is slightly less verbose. (`await` is part of `PlaySpecification`, see `play.api.test.FutureAwaits`).

marcospereira commented on framework/build.sbt:
```diff
@@ -324,14 +325,24 @@ lazy val PlayEhcacheProject = PlayCrossBuiltProject("Play-Ehcache", "play-ehcach
 
+lazy val PlayCaffeineCacheProject = PlayCrossBuiltProject("Play-Caffeine-Cache", "play-caffeine-cache")
+    .settings(
+      libraryDependencies ++= playCaffeineDeps
```
Also add the following setting here:

```scala
mimaPreviousArtifacts := Set.empty,
```

PromanSEW commented on documentation/manual/releases/release27/Highlights27.md:
```diff
@@ -24,3 +24,11 @@ public class MyForm {
+
+## Support for Caffeine
+
+Play now offers a CacheApi implementation based on [Caffeine](https://github.com/ben-manes/caffeine/). Caffeine is the recomended cache implementation for Play users.
```
Typo: recom**m**ended

mkurz commented on framework/src/play-caffeine-cache/src/main/resources/reference.conf:
```diff
@@ -0,0 +1,26 @@
+      defaults {
+        initial-capacity = null
+        weak-keys = null
+        weak-keys = false
```
@dusanstanojeviccs I think you made a mistake here? `weak-keys` twice? Shouldn't be one of both configs be `weak-values`? See https://github.com/playframework/playframework/blob/bc1badd22f36bcfe09517be37a9ef66559b53b31/cache/play-caffeine-cache/src/main/java/play/cache/caffeine/CaffeineParser.java#L68-L70
Also the config `maximum-size = null` is missing: https://github.com/playframework/playframework/blob/bc1badd22f36bcfe09517be37a9ef66559b53b31/cache/play-caffeine-cache/src/main/java/play/cache/caffeine/CaffeineParser.java#L60-L64

mkurz replied:
Fix can be found here: #10398
